x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -687,23 +691,57 @@ private DomainConfig getDomainConfigForSubject(Subject subject) {
         }
     }
 
-    private void updateProfileForActivate(Subject subject, VersionedDocument versionedDocument, ActionListener<Profile> listener)
-        throws IOException {
-        final ProfileDocument profileDocument = updateWithSubject(versionedDocument.doc, subject);
+    // package private for testing
+    void updateProfileForActivate(Subject subject, VersionedDocument currentVersionedDocumentBySearch, ActionListener<Profile> listener) {
+        final ProfileDocument newProfileDocument = updateWithSubject(currentVersionedDocumentBySearch.doc, subject);
 
-        doUpdate(
-            buildUpdateRequest(
-                profileDocument.uid(),
-                wrapProfileDocumentWithoutApplicationData(profileDocument),
-                RefreshPolicy.WAIT_UNTIL,
-                versionedDocument.primaryTerm,
-                versionedDocument.seqNo
-            ),
-            listener.map(
-                updateResponse -> new VersionedDocument(profileDocument, updateResponse.getPrimaryTerm(), updateResponse.getSeqNo())
-                    .toProfile(Set.of())
-            )
-        );
+        if (shouldSkipUpdateForActivate(currentVersionedDocumentBySearch.doc, newProfileDocument)) {
+            listener.onResponse(currentVersionedDocumentBySearch.toProfile(Set.of()));
+            return;
+        }
+
+        // The current document by search may not be up-to-date, using GET to retrieve it again to double check whether
+        // update is truly necessary. Note the document by GET is only used as an optimisation technique for skipping the update.
+        // It does not affect the result of update when the update does need to happen.
+        getVersionedDocument(currentVersionedDocumentBySearch.doc.uid(), ActionListener.wrap(currentVersionedDocumentByGet -> {
+            if (shouldSkipUpdateForActivate(currentVersionedDocumentByGet.doc, newProfileDocument)) {
+                listener.onResponse(currentVersionedDocumentByGet.toProfile(Set.of()));
+            } else {
+                doUpdate(
+                    buildUpdateRequest(
+                        newProfileDocument.uid(),
+                        wrapProfileDocumentWithoutApplicationData(newProfileDocument),
+                        RefreshPolicy.WAIT_UNTIL
+                    ),
+                    listener.map(
+                        updateResponse -> new VersionedDocument(
+                            newProfileDocument,
+                            updateResponse.getPrimaryTerm(),
+                            updateResponse.getSeqNo()
+                        ).toProfile(Set.of())
+                    )
+                );
+            }
+        }, listener::onFailure));
+    }
```

Comment:
This is the gist of this PR. The code does GET before update so it can reliably tell whether the document has been updated already within past 30 seconds (the critical part is really whether it is updated in the past 1 second because the document-by-search cannot reliably tell that).